File: java/eclipse-workspace/anclient.jsample/src/main/java/io/odysz/jclient/Clients.java
package io.odysz.jclient;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.sql.SQLException;

import io.odysz.anson.x.AnsonException;
import io.odysz.common.AESHelper;
import io.odysz.common.Utils;
import io.odysz.semantic.jprotocol.AnsonMsg;
import io.odysz.semantic.jprotocol.AnsonMsg.Port;
import io.odysz.semantic.jprotocol.AnsonResp;
import io.odysz.semantic.jprotocol.IPort;
import io.odysz.semantic.jsession.AnSessionReq;
import io.odysz.semantic.jsession.AnSessionResp;
import io.odysz.semantics.x.SemanticException;

/**
 * @author odys-z@github.com
 * @param <T>
 */
public class Clients {
	public static final boolean console = true;

	public static String servRt;
	/** DB connection ID. same in connects.xml/t/C/id at server side. */
	private static String conn;


	/**Initialize configuration.
	 * @param servRoot
	 */
	public static void init(String servRoot) {
		servRt = servRoot;
		conn = null; // client can't control engine connect. configured in workflow-meta.xml
	}
	
	/**Login and return a client instance (with session managed by jserv).
	 * @param uid
	 * @param pswdPlain
	 * @return null if failed, a SessionClient instance if login succeed.
	 * @throws SQLException the request makes server generate wrong SQL.
	 * @throws SemanticException Request can not parsed correctly 
	 * @throws GeneralSecurityException  other error
	 * @throws Exception, most likely the network failed
	 */
	public static AnsonClient login(String uid, String pswdPlain)
			throws IOException, SemanticException, SQLException, GeneralSecurityException, AnsonException {
		byte[] iv =   AESHelper.getRandom();
		String iv64 = AESHelper.encode64(iv);
		if (uid == null || pswdPlain == null)
			throw new SemanticException("user id and password can not be null.");
		String tk64 = AESHelper.encrypt(uid, pswdPlain, iv);
		
		// formatLogin: {a: "login", logid: logId, pswd: tokenB64, iv: ivB64};
  		// AnsonMsg<? extends AnsonBody> reqv11 = new AnsonMsg<AnQueryReq>(Port.session);;
		AnsonMsg<AnSessionReq> reqv11 = AnSessionReq.formatLogin(uid, tk64, iv64);

		AnsonClient[] inst = new AnsonClient[1]; 

		HttpServClient httpClient = new HttpServClient();
		String url = servUrl(Port.session);
		httpClient.postV11(url, reqv11, (code, msg) -> {
					if (AnsonMsg.MsgCode.ok == code) {
						// create a logged in client
						inst[0] = new AnsonClient(((AnSessionResp) msg).ssInf());

						if (Clients.console)
							Utils.logi(msg.toString());
					}
					else throw new SemanticException(
							"loging failed\ncode: %s\nerror: %s",
							code, ((AnsonResp)msg).msg());
				});
  		if (inst[0] == null)
  			throw new IOException("HttpServClient return null client.");
  		return inst[0];
	}
	
	/**Get a insecure client - for no session handling request.
	 * @param uid
	 * @param pswdPlain
	 * @return a client can only read data from server.
	 * @throws Exception
	public static SessionClient readOnly(String uid, String pswdPlain) throws Exception {
		return new InsecureClient(servRt, conn);
	}
	 */

	/**Helper for generate serv url (with configured server root and db connection ID).
	 * @param port
	 * @return url, e.g. http://localhost:8080/query.serv?conn=null
	 */
	static String servUrl(IPort port) {
		return String.format("%s/%s?conn=%s", servRt, port.url(), conn);
	}
}

File: csharp/anclient/anclient/src/anclient/Clients.cs
using io.odysz.common;
using io.odysz.semantic.jprotocol;
using io.odysz.semantic.jsession;
using io.odysz.semantics.x;
using System;
using static io.odysz.semantic.jprotocol.AnsonMsg;

namespace io.odysz.anclient
{
    public class Clients
    {
		public const bool console = true;

		public static string servRt;

		/// <summary> DB connection ID. same in connects.xml/t/C/id at server side.
		/// </summary>
		private static string conn;


		/// <summary>Initialize configuration.
		/// </summary>
		/// <param>servRoot</param>
		public static void Init(string servRoot)
		{
			servRt = servRoot;
			conn = null; // client can't control engine connect. configured in workflow-meta.xml
		}

		/// <summary>Login and return a client instance (with session managed by jserv).
		/// </summary>
		/// <paramref name="uid"/>
		/// <param name="pswdPlain">password in plain</param>
		/// <return> null if failed, a SessionClient instance if login succeed.</return>
		/// <throws> SQLException the request makes server generate wrong SQL.</throws>
		/// @throws SemanticException Request can not parsed correctly
		/// @throws GeneralSecurityException  other error
		/// @throws Exception, most likely the network failed
		public static AnsonClient Login(string uid, string pswdPlain, Action<int, AnSessionResp> onlogin = null, Action<int, AnsonResp> onerror = null)
		{
            byte[] iv = AESHelper.getRandom();
            string iv64 = AESHelper.Encode64(iv);
            if (uid == null || pswdPlain == null)
                throw new SemanticException("user id and password can not be null.");
            string tk64 = AESHelper.Encrypt(uid, pswdPlain, iv);

            // formatLogin: {a: "login", logid: logId, pswd: tokenB64, iv: ivB64};
            // AnsonMsg<? extends AnsonBody> reqv11 = new AnsonMsg<AnQueryReq>(Port.session);;
            AnsonMsg reqv11 = AnSessionReq.formatLogin(uid, tk64, iv64);

            AnsonClient[] inst = new AnsonClient[1];

            HttpServClient httpClient = new HttpServClient();
            string url = ServUrl(new Port(Port.session));
            httpClient.Post(url, reqv11, (code, msg) => {
                    if (MsgCode.ok == code.code) {
                        // create a logged in client
                        inst[0] = new AnsonClient(((AnSessionResp) msg.Body()[0]).ssInf);
                        if (onlogin != null)
                            onlogin(code.code, (AnSessionResp)msg.Body()[0]);

                            if (Clients.console)
                                Console.WriteLine(msg.ToString());
                }
                                else throw new SemanticException(
                                        "loging failed\ncode: {0}\nerror: {1}",
                                        code, ((AnSessionResp) msg.Body()[0]).Msg());
            } );
            return inst[0];
        }

        /**Helper for generate serv url (with configured server root and db connection ID).
         * @param port
         * @return url, e.g. http://localhost:8080/query.serv?conn=null
         */
        public static string ServUrl(Port port)
        {
            return string.Format("{0:S}/{1:S}?conn={2:S}", servRt, port.Url(), conn);
        }

    }
}
